### ISSUE TITLE
Decoding issue

### ISSUE THREAD

**Issue by mkperez808** [mravanelli/pytorch-kaldi#25]
I'm getting an error in the decoding section of run_exp.py.
When running cmd_decode (kaldi_decoding_scripts/decode_dnn.sh) I'm a getting an ls error saying that the script cannot aches these .ark files. Are these .ark files supposed to be generated somewhere else in the code? Or are these .ark files created after running decode_dnn.sh?

log file output:

> /bin/sh: 1: ./path.sh: not found

1947

hmm-info /z/mkperez/kaldi/egs/timit/s5/exp/dnn4_pretrain-dbn_dnn_ali/final.mdl 

ali-to-pdf /z/mkperez/kaldi/egs/timit/s5/exp/dnn4_pretrain-dbn_dnn_ali/final.mdl 'ark:gunzip -c /z/mkperez/kaldi/egs/timit/s5/exp/dnn4_pretrain-dbn_dnn_ali/ali.*.gz |' ark:- 
LOG (ali-to-pdf[5.0.13~1-95f0f]:main():ali-to-pdf.cc:68) Converted 5352 alignments to pdf sequences.
LOG (analyze-counts[5.0.13~1-95f0f]:main():analyze-counts.cc:193) Summed 5352 int32 vectors to counts, skipped 0 vectors.
LOG (analyze-counts[5.0.13~1-95f0f]:main():analyze-counts.cc:195) Counts written to exp/TIMIT_liGRU_mfcc/exp_files/forward_out_dnn2_lab_cd.count

kaldi_decoding_scripts/decode_dnn.sh /home/mkperez/pytorch-kaldi/exp/TIMIT_liGRU_mfcc/decoding_TIMIT_test_out_dnn2.conf /home/mkperez/pytorch-kaldi/exp/TIMIT_liGRU_mfcc/decode_TIMIT_test_out_dnn2 /home/mkperez/pytorch-kaldi/exp/TIMIT_liGRU_mfcc/exp_files/forward_TIMIT_test_ep*_ck*_out_dnn2_to_decode.ark
featString /home/mkperez/pytorch-kaldi/exp/TIMIT_liGRU_mfcc/exp_files/forward_TIMIT_test_ep*_ck*_out_dnn2_to_decode.ark 
kaldi_decoding_scripts/decode_dnn.sh: no such file /home/mkperez/pytorch-kaldi/quick_test/graph/HCLG.fst

ls: cannot access '/home/mkperez/pytorch-kaldi/exp/TIMIT_liGRU_mfcc/exp_files/forward_TIMIT_test_ep*_ck*_out_dnn2_to_decode.ark': No such file or directory

>

**Comment by TParcollet** [comment]
Does this "/home/mkperez/pytorch-kaldi/exp/TIMIT_liGRU_mfcc/exp_files/forward_TIMIT_test_ep*_ck*_out_dnn2_to_decode.ark" file exists ? 

If not, please make sure that you don't have any error at the end of the training procedure ?

**Comment by mkperez808** [comment]
These files don't exist. Where exactly in the training procedure should these files be produced? My log file doesn't seem to have any errors indicating failure. 

Log file:
copy-feats scp:exp/TIMIT_MLP_basic/exp_files/train_TIMIT_tr_ep005_ck02_mfcc.lst ark:- 
add-deltas --delta-order=2 ark:- ark:- 
apply-cmvn --utt2spk=ark:/z/mkperez/kaldi/egs/timit/s5/data/train/utt2spk ark:/z/mkperez/kaldi/egs/timit/s5/mfcc/cmvn_train.ark ark:- ark:- 
LOG (copy-feats[5.0.13~1-95f0f]:main():copy-feats.cc:120) Copied 740 feature matrices.
LOG (apply-cmvn[5.0.13~1-95f0f]:main():apply-cmvn.cc:146) Applied cepstral mean normalization to 740 utterances, errors on 0
ali-to-pdf /z/mkperez/kaldi/egs/timit/s5/exp/dnn4_pretrain-dbn_dnn_ali/final.mdl ark:- ark:- 
LOG (ali-to-pdf[5.0.13~1-95f0f]:main():ali-to-pdf.cc:68) Converted 5352 alignments to pdf sequences.

Where exactly should these .ark files be created in the training process?

**Comment by mravanelli** [comment]
The *.ark file contains the posterior probabilities on the test data, that
will be given in input to the decoder. This file is created during the
forward step that happens after the training step. Could you post the
res.res file?

**Comment by mkperez808** [comment]
Thanks for the speedy reply. I forgot to add the corresponding paths for the "lab" section of the dataset in the cfg file. I am now getting an error when decoding the log file in the decoding directory is:

> latgen-faster-mapped --min-active=200 --max-active=7000 --max-mem=50000000 --beam=13.0 --lattice-beam=8.0 --acoustic-scale=0.2 --allow-partial=true --word-symbol-table=/z/mkperez/kaldi/egs/timit/s5/exp/tri3/graph/words.txt /z/mkperez/kaldi/egs/timit/s5/exp/dnn4_pretrain-dbn_dnn_ali_test/final.mdl /z/mkperez/kaldi/egs/timit/s5/exp/tri3/graph/HCLG.fst 'ark,s,cs: cat /home/mkperez/pytorch-kaldi/exp/TIMIT_MLP_basic/exp_files/forward_TIMIT_test_ep023_ck00_out_dnn1_to_decode.ark |' 'ark:|gzip -c > /home/mkperez/pytorch-kaldi/exp/TIMIT_MLP_basic/decode_TIMIT_test_out_dnn1/lat.1.gz' 
WARNING (latgen-faster-mapped[5.0.13~1-95f0f]:OpenInternal():kaldi-io.cc:792) Invalid input filename format ' cat /home/mkperez/pytorch-kaldi/exp/TIMIT_MLP_basic/exp_files/forward_TIMIT_test_ep023_ck00_out_dnn1_to_decode.ark |'
WARNING (latgen-faster-mapped[5.0.13~1-95f0f]:Open():util/kaldi-table-inl.h:513) Failed to open stream ' cat /home/mkperez/pytorch-kaldi/exp/TIMIT_MLP_basic/exp_files/forward_TIMIT_test_ep023_ck00_out_dnn1_to_decode.ark |'
ERROR (latgen-faster-mapped[5.0.13~1-95f0f]:SequentialTableReader():util/kaldi-table-inl.h:876) Error constructing TableReader: rspecifier is ark,s,cs: cat /home/mkperez/pytorch-kaldi/exp/TIMIT_MLP_basic/exp_files/forward_TIMIT_test_ep023_ck00_out_dnn1_to_decode.ark |

[ Stack-Trace: ]
latgen-faster-mapped() [0x84fb9e]
kaldi::MessageLogger::HandleMessage(kaldi::LogMessageEnvelope const&, char const*)
kaldi::MessageLogger::~MessageLogger()
kaldi::SequentialTableReader<kaldi::KaldiObjectHolder<kaldi::Matrix<float> > >::SequentialTableReader(std::string const&)
main
__libc_start_main
_start

>

**Comment by mkperez808** [comment]
The script runs if I delete the extra space in decode_dnn.sh:
I changed the line:` finalfeats="ark,s,cs: cat $ck_data |"`
to: `finalfeats="ark,s,cs:cat $ck_data |"`

also, in run_exp.py in the decoding section of the code. When I set the forward_save_files to True the generated the _to_decode.ark files get deleted. 
Suggested change: `if not forward_save_files:`

Thank you so much for creating this tool. It's super helpful!
